### PR TITLE
Fix group for equal sized systems

### PR DESCRIPTION
**Pull request by nfwvogt**

### CODE REVIEW

File: struqture/src/mixed_systems/mixed_open_system.rs [hunk 1]
```diff
@@ -114,6 +114,177 @@ impl<'a> OpenSystem<'a> for MixedLindbladOpenSystem {
...
+        for (index, (system_spins, noise_spins)) in current_number_spins
+            .iter_mut()
+            .zip(noise_number_spins.iter())
+            .enumerate()
+        {
+            if system_spins != noise_spins {
+                match (*system_spins, noise_spins) {
+                    (Some(n), None) => {
+                        if n < noise_number_current_spins[index] {
+                            return Err(StruqtureError::MissmatchedNumberSubsystems {
```
Comment (rreiner-hqs) [inline struqture/src/mixed_systems/mixed_open_system.rs:131]: It looks like here the error is raised if the number of spins in a particular subsystem are missmatched, but the error just shows the number of subsystems of each type, which might as well be the same between the system and noise parts

File: struqture/tests/fermions/fermionic_open_system.rs [hunk 2]
```diff
@@ -62,11 +62,38 @@ fn group() {
     assert_eq!(slos, FermionLindbladOpenSystem::default())
 }
 
+#[test]
+fn group_with_none() {
+    let slos = FermionLindbladOpenSystem::group(
```
Comment (rreiner-hqs) [inline struqture/tests/fermions/fermionic_open_system.rs:67]: Probably not important, but maybe we want to call this `flos` instead

File: struqture/tests/mixed_systems/mixed_open_system.rs [hunk 3]
```diff
@@ -86,11 +86,54 @@ fn group() {
...
+    );
+    assert!(slos.is_err());
+
+    let slos = MixedLindbladOpenSystem::group(
+        MixedHamiltonianSystem::new([None], [Some(3)], [None]),
+        MixedLindbladNoiseSystem::new([Some(2)], [Some(2)], [Some(2)]),
+    );
+    assert!(slos.is_err());
+    let slos = MixedLindbladOpenSystem::group(
+        MixedHamiltonianSystem::new([None], [None], [Some(3)]),
```
Comment (rreiner-hqs) [inline struqture/tests/mixed_systems/mixed_open_system.rs:134]: This is the same as the first case, we probably want `MixedHamiltonianSystem::new([Some(3)], [None], [None])`

File: struqture/src/mixed_systems/mixed_open_system.rs [hunk 4]
```diff
@@ -114,6 +114,177 @@ impl<'a> OpenSystem<'a> for MixedLindbladOpenSystem {
                 actual_number_fermion_subsystems: noise.number_fermions.len(),
             });
         }
+
+        let mut current_number_spins = system.number_spins.clone();
+        let noise_number_spins = noise.number_spins.clone();
+        let noise_number_current_spins = noise.number_spins();
+        let system_number_current_spins = system.number_spins();
```
Comment (kbarkhqs) [inline struqture/src/mixed_systems/mixed_open_system.rs:121]: The naming here doesn't seem to be consistent. Also, for line 118 and line 120, based on the name, should the `current_number_spins` method be called?

Comment (nfwvogt) [reply struqture/src/mixed_systems/mixed_open_system.rs:121]: Fixed

File: struqture/src/mixed_systems/mixed_open_system.rs [hunk 5]
```diff
@@ -114,6 +114,177 @@ impl<'a> OpenSystem<'a> for MixedLindbladOpenSystem {
...
+                    _ => panic!("Unexpected missmatch of number modes"),
+                }
+            }
+        }
+        let mut system = system;
+        let mut noise = noise;
+        system.number_spins = current_number_spins.clone();
+        noise.number_spins = current_number_spins;
+
+        // Checking boson compatability
```
[the remark this answers is not in the document]
Comment (nfwvogt) [reply struqture/src/mixed_systems/mixed_open_system.rs:174]: Fixed

File: struqture/src/mixed_systems/mixed_open_system.rs [hunk 6]
```diff
@@ -114,6 +114,177 @@ impl<'a> OpenSystem<'a> for MixedLindbladOpenSystem {
...
+                    }
+                    _ => panic!("Unexpected missmatch of number modes"),
+                }
+            }
+        }
+
+        system.number_bosons = current_number_bosons.clone();
+        noise.number_bosons = current_number_bosons;
+
+        // Checking Fermion compatability
```
[the remark this answers is not in the document]
Comment (nfwvogt) [reply struqture/src/mixed_systems/mixed_open_system.rs:231]: Fixed

File: struqture/tests/bosons/bosonic_open_system.rs [hunk 7]
```diff
@@ -57,11 +57,38 @@ fn group() {
     assert_eq!(slos, BosonLindbladOpenSystem::default())
 }
 
+#[test]
+fn group_with_none() {
+    let slos = BosonLindbladOpenSystem::group(
```
[the remark this answers is not in the document]
Comment (nfwvogt) [reply struqture/tests/bosons/bosonic_open_system.rs:62]: Fixed

File: struqture/tests/fermions/fermionic_open_system.rs (hunk 2 above)
Comment (nfwvogt) [reply struqture/tests/fermions/fermionic_open_system.rs:67]: Fixed

File: struqture/tests/mixed_systems/mixed_open_system.rs [hunk 8]
```diff
@@ -86,11 +86,54 @@ fn group() {
 
 // Test the group function of the MixedLindbladOpenSystem
 #[test]
-fn group_failing() {
+fn group_with_none() {
     let slos = MixedLindbladOpenSystem::group(
```
[the remark this answers is not in the document]
Comment (nfwvogt) [reply struqture/tests/mixed_systems/mixed_open_system.rs:90]: Fixed

File: struqture/tests/mixed_systems/mixed_open_system.rs (hunk 3 above)
Comment (nfwvogt) [reply struqture/tests/mixed_systems/mixed_open_system.rs:134]: Fixed

File: struqture/src/mixed_systems/mixed_open_system.rs (hunk 1 above)
Comment (nfwvogt) [reply struqture/src/mixed_systems/mixed_open_system.rs:131]: Fixed

**Review by kbarkhqs — APPROVED** [review]

**Pull request merged**